src/d_soundfile.c

```diff
@@ -1209,25 +1210,284 @@ static void soundfile_xferout_words(int nchannels, t_word **vecs,
+                    x->x_vecs[i][j].w_float = 0;
+            }
+            fclose(fp);
+
+        lost:
+
+            if (x->x_fd >= 0)
+            {
+                close (x->x_fd);
+                x->x_fd = -1;
+            }
+            pthread_mutex_lock(&x->x_mutex);
+            x->x_requestcode = SF_REQUEST_NOTHING;
+            sf_cond_signal(&x->x_answercondition);
+        }
+        else if (x->x_requestcode == SF_REQUEST_QUIT)
+        {
+            if (x->x_fd >= 0)
+            {
+                pthread_mutex_unlock(&x->x_mutex);
+                close (x->x_fd);
+                pthread_mutex_lock(&x->x_mutex);
+                x->x_fd = -1;
+            }
+            soundfiler_child_cleangarbage(x);
+            x->x_requestcode = SF_REQUEST_NOTHING;
+            sf_cond_signal(&x->x_answercondition);
+            break;
+        }
+    }
+    pthread_mutex_unlock(&x->x_mutex);
+    return (0);
+}
+
+static void soundfiler_tick(t_soundfiler *x)
+{
+    int i;
+    t_threaded_data *data = x->x_data;
+    
+    if (data->x_state == SF_STATE_IDLE) return; /* no pending operation */
+
+    if (data->x_state != SF_STATE_READ) return; /* only read for now */
+
+    pthread_mutex_lock(&data->x_mutex);
+    if (data->x_requestcode != SF_REQUEST_NOTHING) /* thread is still reading */
+    {
+        sf_cond_signal(&data->x_requestcondition);
+        pthread_mutex_unlock(&data->x_mutex);
+        clock_delay(data->x_clock, SOUNDFILER_GRAIN); /* retry later */
+        return;
+    }
+
+    pthread_mutex_unlock(&data->x_mutex);
+        /* should never happen: */
+    if (data->x_ngarbages) pd_error(x, "soundfiler: remaining garbage, resulting memory leak!"); 
+    data->x_ngarbages = 0;
+
+    for (i = 0; i < data->x_narrays; i++)
+    {
+        int vecsize;
+        t_word *vec;
+        t_garray *garray;
+        t_array *array;
+        char *oldvec;
+        int oldvecsize; 
+
+        if (!(garray = (t_garray *)pd_findbyclass(data->x_arrays[i], garray_class)))
+        {
+            pd_error(x, "%s: no such table", data->x_arrays[i]->s_name);
+            if (data->x_vecs[i]) {
+                freebytes(data->x_vecs[i], data->x_vecsize);
+                data->x_vecs[i] = 0;
+            }
+            continue;
+        }
+        else if (!garray_getfloatwords(garray, &vecsize, &vec))
+            error("%s: bad template", data->x_arrays[i]->s_name);
+
+        if (!data->x_vecs[i]) continue;
+
+        if (!(array = garray_getarray(garray))) 
+            error("%s: bad template", data->x_arrays[i]->s_name);
+
+        oldvec = array->a_vec;
+        oldvecsize = array->a_n;
+        array->a_vec = (char *)data->x_vecs[i];
+        array->a_n = data->x_vecsize;
+        garray_resize_long(garray, data->x_vecsize); /* fake resize, to resync GUI and DSP */
+            /* for sanity's sake let's clear the save-in-patch flag here */
+        garray_setsaveit(garray, 0);
```

Comment:
I just took this from the original soundfiler code, as I tried to keep the behavior the most I could.

Reply:
I see! In the original [soundfiler] code this is only then when using "-resize", probably to avoid issues like these: https://github.com/pure-data/pure-data/issues/567

Actually, I'm working on a fix to make this obsolete: https://github.com/pure-data/pure-data/pull/621

For now, you could add a "x_resize" member to the thread data and only call `garray_setsaveit` if necessary.

Reply:
x_resize already exists; I did what you said.